### ISSUE TITLE
[HOLD for payment 2024-04-01] [$250] Android - App lags after adding code block in task description

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#34324]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.24-0
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** Applause internal team
**Slack conversation:**

## Action Performed:
1. Open the app and log in
2. Tap the FAB > Assign Task
3. Enter title
4. Add the code block in the task description:
if (files.length > 0) {
// Prevent the default so we do not post the file name into the text box
event.preventDefault(); this.props.onPasteFile(event.clipboardData.files[0]);
return;
if (files.length > 0) {
// Prevent the default so we do not post the file name into the text box
event.preventDefault() this.props.onPasteFile(event.clipboardData.files[0]);
return;
5. Proceed to the next screen and finish creating the task
## Expected Result:
Able to create a task without significant delay

## Actual Result:
The app lags after pasting the code block in the description field

## Workaround:
unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [ ] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

https://github.com/Expensify/App/assets/43996225/24998581-9572-4d15-b18f-d5cacaa98ff7

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~013b44c7d5e9e29617</li>
        <li>Upwork Job ID: 1745374242762850304</li>
        <li>Last Price Increase: 2024-03-15</li>
<li>Automatic offers: </li>
<ul>
<li>ntdiary | Reviewer | 0</li>
<li>aswin-s | Contributor | 0</li>
</ul></li>
    </ul>
</details>

**Comment by garrettmknight** [comment]
Waiting on proposals

**Comment by sfurqan2** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
The app lags when code is entered in description textbox
### What is the root cause of that problem?
From what I have investigated so far, I think default value in the component is parsing the text entered into the textbox. When we enter the code above, I think the parser takes longer time to render which causes the lag.
https://github.com/Expensify/App/blob/8e2fb675389bb84deee4b8e001a536252883e72f/src/pages/tasks/NewTaskDetailsPage.js#L116
### What changes do you think we should make in order to solve the problem?
I think we should remove parser from default value attribute in text input and move it to where "task description" state default value is set.


### What alternative solutions did you explore? (Optional)
NA

**Comment by aswin-s** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Adding code block in task description makes the page laggy.

### What is the root cause of that problem?

PR #29144 modified the task description input logic to support mark down. Task description is parsed to mark down before saving and vice versa when setting the default value to input.

However the parsing logic which the method `parser.replace()` is computationally intensive. 

![image](https://github.com/Expensify/App/assets/6880914/083fe4a2-1526-4694-a6bf-057ddf39e2f1)

CPU Profiling above shows `modifyTextForUrlLinks`, which gets invoked by `parser.replace` taking around 6 seconds to execute when transitioning from New Task to Confirm task page.  

There are 3 locations in Task creation where `parser.replace` is used.

https://github.com/Expensify/App/blob/8ba2782c8c35a394a3e1f8d8406085d7e1b49215/src/pages/tasks/NewTaskDetailsPage.js#L116

https://github.com/Expensify/App/blob/8ba2782c8c35a394a3e1f8d8406085d7e1b49215/src/pages/tasks/NewTaskDescriptionPage.js#L72

https://github.com/Expensify/App/blob/8ba2782c8c35a394a3e1f8d8406085d7e1b49215/src/pages/tasks/TaskDescriptionPage.js#L120

Note that `defaultValue` prop is directly assigned a function call instead of a memoized value. This causes the computationally intensive method to be invoked whenever the page state changes. This in turn causes the JS bridge to get blocked and animations to lag.


Also the useEffect here is dependent on `props.task`. 

https://github.com/Expensify/App/blob/d4c28a4b95eb0a744a26fc2756fc2a42bb744f12/src/pages/tasks/NewTaskDetailsPage.js#L46-L49

When navigating from NewTaskDetailPage to NewTaskPage(ConfirmTask) screen , the first screen is not unmounted by ReactNavigation. It keeps the page in memory for navigating back. This causes the useEffect to get executed when task object is updated in Onyx. Which also invokes the parser.replace method.

So overall a computationally intensive method gets unnecessarily executed multiple times causing the app to lag.

### What changes do you think we should make in order to solve the problem?

Memoize the defaultValue prop using React.useMemo, so that parser.replace gets executed only when description changes.

```tsx

const defaultValue = useMemo(() => parser.htmlToMarkdown(parser.replace(taskDescription)), [taskDescription]);

```

Also the useEffect should get executed only when the screen is focused. Update the logic to prevent unwanted execution offscreen.

```tsx

const isFocused = useIsFocused();

useEffect(() => {
  // Update state only if screen is focused
  if (isFocused) {        
    setTaskTitle(props.task.title);
    setTaskDescription(parser.htmlToMarkdown(parser.replace(props.task.description || '')));
  }
}, [props.task.title, props.task.description, isFocused]);


```


### What alternative solutions did you explore? (Optional)

The other area to focus on is actually optimising the parser.replace logic in ExpensiMark library. However it is in Expensify-Common repository and possibly shared by multiple apps.

**Comment by sourcecodedeveloper** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Adding inlinecode block in task description is considered valid input and lags the page.

### What is the root cause of that problem?
The root cause of the issue is the task description considers inline code valid input. When html markup text is entered in the task description, the validation raises error and form cannot be submited. 

The formprovider checks for html tag index and raises validation error should the task description have any html tags

https://github.com/Expensify/App/blob/a8acf44090c9aab4e9af7967a79758472ada8fc3/src/components/Form/FormProvider.js#L137

### What changes do you think we should make in order to solve the problem?

To keep the solution simple, the html tag check is already being done, another regex can be made to check for inline code and raise validation error.

In the const.ts add regex to test for common inline code patterns
```
VALIDATE_FOR_INLINE_CODEBLOCK_REGEX: /(this.)|(\[\d\])|(\(\);)|(\(.*\))/g,
```

In FormProvider, test the inputValue,
```

// Check for inline code
const foundInlineCodeIndex = inputValue.search(CONST.VALIDATE_FOR_INLINE_CODEBLOCK_REGEX);

// Return early if there are no HTML characters
if (leadingSpaceIndex === -1 && foundHtmlTagIndex === -1 && foundInlineCodeIndex === -1 ) {
    return;
}
```

The form would be prevented from submitted and when the text description is valid, the ExpensiMark parser.replace() is not causing any lag.



### What alternative solutions did you explore? (Optional)

**Comment by garrettmknight** [comment]
@ntdiary looks like we've got some proposals for this one. Can you take a look when you get a chance?

**Comment by sourcecodedeveloper** [comment]
@ntdiary The parser.replace only lags when inline code is entered otherwise there is no lag. The regex in FormProvider considers html tags as invalid unless in allowed tags list. The requirements needs to be cleared whether inlinecode should be considered valid or invalid input.

**Comment by sfurqan2** [comment]
Hey @ntdiary, I was of the assumption that we only show inline code on page load, as I did not see other components parsing the updated description text. Please let me know if that's not the correct behavior.

**Comment by sfurqan2** [comment]
I agree then, we must think of some way to optimize this regex. :)

**Comment by aswin-s** [comment]
@ntdiary The real culprit is the [huge tlds regex](https://raw.githubusercontent.com/Expensify/expensify-common/main/lib/tlds.jsx) that's being used in `MARKDOWN_URL_REGEX`. To test the theory I replaced the `TLD_REGEX` const with a much simpler regex which simply validates if the string is a valid `tlds` instead of explicitly matching specific TLDs. 

```tsx
const TLD_REGEX='(?:[a-zA-Z]{2,63}|xn--[a-zA-Z0-9-]{4,59})';
```

And here are the results for a single execution of parser.replace(). With original TLD_REGEX execution time is approximately **1s** where as for simpler regex it is **10ms**. Also it is important to note that the execution time is directly proportional to the input length. If I copy over the same test input 4 times, execution time with original regex becomes 4 seconds. That's 1s for around 350 characters. Which is quite huge!


| Before | After |
|--|--|
|![before](https://github.com/Expensify/App/assets/6880914/80a01330-da1a-4e5b-a0da-c37cb31d3853) |![after](https://github.com/Expensify/App/assets/6880914/70f4ba05-c8a5-4ea3-83e7-efc84fb2479f) |


So it boils down to the decision whether we can trim down the specific TLD names with in the TLD_REGEX.

**Comment by garrettmknight** [comment]
@ntdiary any thoughts on the suggestions?

**Comment by aswin-s** [comment]
@ntdiary I was not proposing to redefine TLD_REGEX as `(?:[a-zA-Z]{2,63}|xn--[a-zA-Z0-9-]{4,59})`. Just pointing out that usage of a smaller regex significantly reduces execution time. As it stands I don't have a solid proposal at the moment without considerably changing the auto linking patterns.

However I do like to point out that we do have a very liberal definition for what a valid url is. Any set of words separated by dots and ending with a valid TLD is considered as a valid url as per MARKDOWN_LINK_REGEX. That's why we need to rely on explicit TLD names to parse links in markdown. Because most set of words pass through the initial set of conditionals only to fail ultimately at the specific TLD check. That's why the simplified regex matched any two words separated by dots as links. 

Also the issue seems to impact only Android & iOS platforms. mWeb seems to be largely unaffected. So I suspect Hermes to be the culprit.

**Comment by sourcecodedeveloper** [comment]
> A simple fix would be to split the hyphen check into a subsequent conditional validation, but I need to ask on Slack to see if this is appropriate.


@ntdiary The below regex adds conditional validation to URL_WEBSITE_REGEX for whether to use the double hypen TLD_REGEX or single hypen regex URL_HYPEN_REGEX

(?(?=[-]{2})(?:${TLD_REGEX})|(?:${URL_HYPEN_REGEX})+)

The ?= regex checks for double hypens i.e XN--NQV7FS00EMA|XN--MGBC0A9AZCG and if so run the TLD_REGEX otherwise run the single hypen regex URL_HYPEN_REGEX.

The URL_WEBSITE_REGEX with conditional validation
const URL_WEBSITE_REGEX = `${URL_PROTOCOL_REGEX}?((?:www\\.)?[a-z0-9](?(?=[-]{2})(?:${TLD_REGEX})|(?:${URL_HYPEN_REGEX})+)(?:\\:${ALLOWED_PORTS}|\\b|(?=_))(?!@(?:[a-z\\d-]+\\.)+[a-z]{2,})`;

But need to confirm if condition validation would be acceptable approach.

**Comment by sourcecodedeveloper** [comment]
> The hyphen check I mentioned refers to the changes introduced in PR https://github.com/Expensify/expensify-common/pull/523.
> The simple fix I mentioned is to revert their PR, and then in subsequent functions (such as modifyTextForUrlLinks), validate that if the matched url has an invalid hyphen format (e.g., -abc.com, abc-.com), then do not perform the replacement.

To test the theory, I replaced the URL_WEBSITE_REGEX regex to prior to their PR.
const URL_WEBSITE_REGEX = `(https?:\\/\\/)?((?:www\\.)?[-a-z0-9]+?\\.)+(?:${TLD_REGEX})(?:\\:\\d{2,4}|\\b|(?=_))`;
To match invalid hypen format, I added the regex
const URL_SINGLE_HYPEN_REGEX = `(?:(^-[a-z0-9]+\.[a-z]{3}))|(?:([a-z0-9]+\-\.[a-z]{3}))`; 

Before replacing the URL_WEBSITE_REGEX Regex
LOG  rule name autolink took 2397.42484100163 milliseconds.

After replacing the URL_WEBSITE_REGEX and adding single hypen regex
LOG  autolink regex took 22.16117400303483 milliseconds.

The performance improvement is huge.
@ntdiary Reverting their PR does seem to be the correct approach.

**Comment by sourcecodedeveloper** [comment]
@ntdiary I have revised the [proposal](https://github.com/Expensify/App/issues/34324#issuecomment-1895719504) and included the approach to revert the identified PR and also included POC.

**Comment by garrettmknight** [comment]
@ntdiary when do you think you'll get a chance to review more deeply?

**Comment by sourcecodedeveloper** [comment]
> I need some time to understand your hyphen check to ensure there won't be any regressions. :)

@ntdiary Reverting the PR and then validating the text with hypen regex in the modifyTextForUrlLinks as explained in the comment https://github.com/Expensify/App/issues/34324#issuecomment-1907740798, improves the loading significantly.

const URL_SINGLE_HYPEN_REGEX = `(?:(^-[a-z0-9]+\.[a-z]{3}))|(?:([a-z0-9]+\-\.[a-z]{3}))`; 

Before replacing the URL_WEBSITE_REGEX Regex
LOG rule name autolink took 2397.42484100163 milliseconds.

After replacing the URL_WEBSITE_REGEX and adding single hypen regex
LOG autolink regex took 22.16117400303483 milliseconds.

Reverting the PR would be step 1 and then the URL_WEBSITE_REGEX can be replaced with
```
((?:www\\.)?[a-z0-9](?=(?<label>[-a-z0-9]*[a-z0-9]))\\k<label>?\\.)+(?:${TLD_REGEX})
```

@garrettmknight @ntdiary 
I am open to splitting the bounty, given the regex in the issue is complicated and can use more reviewers for testing.

**Comment by sourcecodedeveloper** [comment]
@ntdiary the ((?:www\\.)?[a-z0-9](?=(?<label>[-a-z0-9]*[a-z0-9]))\\k<label>?\\.)+(?:${TLD_REGEX}) could be used along with early return in case of match. I suggest splitting the bounty and get more reviewers as the regex requires more thorough testing.

**Comment by aswin-s** [comment]
@ntdiary I can confirm that the new pattern fixes the issue and also passes all test cases on `expensify-commons`. 

```js
((?:www\\.)?[a-z0-9](?=(?<label>[-a-z0-9]*[a-z0-9]))\\k<label>?\\.)+(?:${TLD_REGEX})
```

![image](https://github.com/Expensify/App/assets/6880914/4d036bcc-211a-4523-a98c-221b66fabc7f)

**Comment by garrettmknight** [comment]
@aswin-s do you want to take a swing at updating your proposal before we take this one internal?

**Comment by aswin-s** [comment]
@garrettmknight I've [updated my proposal](https://github.com/Expensify/App/issues/34324#issuecomment-1900344893) with the regex modification in expensify-commons.

**Comment by garrettmknight** [comment]
Awesome, thanks! @ntdiary mind giving that another review?

**Comment by garrettmknight** [comment]
Ah, my mistake - @AndrewGable looks like you're up.

**Comment by garrettmknight** [comment]
Andrew is out for a few days - I don't think we need to rush fixing this one so I think we can wait on him returning for review.

**Comment by aswin-s** [comment]
Raised first PR in expensify-commons repo : https://github.com/Expensify/expensify-common/pull/651.

**Comment by AndrewGable** [comment]
Merged Expensify common PR

**Comment by aswin-s** [comment]
Raised followup PR in Expensify/App repo: https://github.com/Expensify/App/pull/36819

**Comment by aswin-s** [comment]
@ntdiary @garrettmknight There is a regression caused by above PR. Should we revert both PRs or just raise a fix. `expensify-commons` have multiple PRs merged after this.

regarding the issue, the below regex doesn't seem to match single character domains like `www.x.com` 

```tsx
((?:www\\.)?[a-z0-9](?=(?<label>[-a-z0-9]*[a-z0-9]))\\k<label>?\\.)+(?:${TLD_REGEX})
```

Adding a matcher for single character domains seems to fix the issue

```tsx
((?:www\\.)?[a-z0-9](?=(?<label>([-a-z0-9]*[a-z0-9])|([a-z0-9]*)))\\k<label>?\\.)+\
(?:${TLD_REGEX})
```

**Comment by garrettmknight** [comment]
@aswin-s how's that PR coming along?

**Comment by garrettmknight** [comment]
Dropping price for regression

**Comment by garrettmknight** [comment]
PR is merged, awaiting deploy to test.

**Comment by garrettmknight** [comment]
@AndrewGable for expensify-common changes, when do we expect to be able to test them? Is it immediately?

**Comment by AndrewGable** [comment]
There needs to be an App PR to use the latest changes of expensify-common, then we can test

**Comment by garrettmknight** [comment]
Great - I'll test again on Monday!

**Comment by garrettmknight** [comment]
Tested and confirmed it's much faster. Setting for payment.

**Comment by garrettmknight** [comment]
**Payment Summary:**
- Contributor: @aswin-s $250 paid via Upwork
- Reviewer: @ntdiary $250 ([link](https://www.upwork.com/nx/wm/offer/101613055))
